python/pyarrow/_array.pyx

```diff
@@ -64,6 +64,56 @@ cdef class DataType:
         else:
             raise TypeError('Invalid comparison')
 
+    def to_pandas_dtype(self):
+        """
+        Return the NumPy dtype that would be used for storing this
+        """
+        cdef Type type_id = self.type.id()
+
+        if type_id == _Type_NA:
+            # NaNs
+            return np.float64
+        elif type_id == _Type_BOOL:
+            return np.bool_
+        elif type_id == _Type_INT8:
+            return np.int8
+        elif type_id == _Type_INT16:
+            return np.int16
+        elif type_id == _Type_INT32:
+            return np.int32
+        elif type_id == _Type_INT64:
+            return np.int64
+        elif type_id == _Type_UINT8:
+            return np.uint8
+        elif type_id == _Type_UINT16:
+            return np.uint16
+        elif type_id == _Type_UINT32:
+            return np.uint32
+        elif type_id == _Type_UINT64:
+            return np.uint64
+        elif type_id == _Type_HALF_FLOAT:
+            return np.float16
+        elif type_id == _Type_FLOAT:
+            return np.float32
+        elif type_id == _Type_DOUBLE:
+            return np.float64
+        elif type_id == _Type_DATE32:
+            return np.dtype('datetime64[ns]')
+        elif type_id == _Type_DATE64:
+            return np.dtype('datetime64[ns]')
+        elif type_id == _Type_TIMESTAMP:
+            return np.dtype('datetime64[ns]')
+        elif type_id == _Type_BINARY:
+            return np.object_
+        elif type_id == _Type_FIXED_SIZE_BINARY:
+            return np.object_
+        elif type_id == _Type_STRING:
+            return np.object_
+        elif type_id == _Type_LIST:
+            return np.object_
+        else:
+            raise NotImplementedError(str(self))
```

Comment:
I've found that having this sort of conversion available as a data structure can be helpful for others when mapping between type systems.  You might consider instead creating a dictionary at the top level and then indexing into it:

```python
numpy_dtype_mapping = {
    _Type_NA: np.float64,
    _Type_BOOL: np.bool_,
    _Type_INT8: np.int8,
    ...
}

def to_pandas_dtype(self):
    return numpy_dtype_mapping[self.type.id()]
```

Although there might be other reasons to avoid this.  I'm not familiar enough with the Arrow codebase to make a judgment here.